Signal/src/view controllers/SignalAttachment.swift

```diff
@@ -0,0 +1,501 @@
+    // for Signal attachments.
+    private class var outputImageUTISet: Set<String> {
+        if allowArbitraryAttachments {
+            return MIMETypeUtil.supportedImageUTITypes()
+        } else {
+            // Until Android client can handle arbitrary attachments,
+            // restrict output.
+            return [
+                kUTTypeJPEG as String,
+                kUTTypeGIF as String,
+                kUTTypePNG as String
+            ]
+        }
+    }
+
+    // Returns the set of UTIs that correspond to valid animated image formats
+    // for Signal attachments.
+    private class var animatedImageUTISet: Set<String> {
+        return MIMETypeUtil.supportedAnimatedImageUTITypes()
+    }
+
+    // Returns the set of UTIs that correspond to valid video formats
+    // for Signal attachments.
+    private class var videoUTISet: Set<String> {
+        if allowArbitraryAttachments {
+            return MIMETypeUtil.supportedVideoUTITypes()
+        } else {
+            return [
+                kUTTypeMPEG4 as String
+            ]
+        }
+    }
+
+    // Returns the set of UTIs that correspond to valid audio formats
+    // for Signal attachments.
+    private class var audioUTISet: Set<String> {
+        if allowArbitraryAttachments {
+            return MIMETypeUtil.supportedAudioUTITypes()
+        } else {
+            return [
+                kUTTypeMP3 as String,
+                kUTTypeMPEG4Audio as String
+            ]
+        }
+    }
+
+    // Returns the set of UTIs that correspond to valid input formats
+    // for Signal attachments.
+    public class var validInputUTISet: Set<String> {
+        return inputImageUTISet.union(videoUTISet.union(audioUTISet))
+    }
+
+    public var isImage: Bool {
+        return SignalAttachment.outputImageUTISet.contains(dataUTI)
+    }
+
+    public var isVideo: Bool {
+        return SignalAttachment.videoUTISet.contains(dataUTI)
+    }
+
+    public var isAudio: Bool {
+        return SignalAttachment.audioUTISet.contains(dataUTI)
+    }
+
+    // Returns an attachment from the pasteboard, or nil if no attachment
+    // can be found.
+    //
+    // NOTE: The attachment returned by this method may not be valid.
+    //       Check the attachment's error property.
+    public class func attachmentFromPasteboard() -> SignalAttachment? {
+        guard UIPasteboard.general.numberOfItems >= 1 else {
+            return nil
+        }
+        // If pasteboard contains multiple items, use only the first.
+        let itemSet = IndexSet(integer:0)
+        guard let pasteboardUTITypes = UIPasteboard.general.types(forItemSet:itemSet) else {
+            return nil
+        }
+        let pasteboardUTISet = Set<String>(pasteboardUTITypes[0])
+        for dataUTI in inputImageUTISet {
+            if pasteboardUTISet.contains(dataUTI) {
+                let data = UIPasteboard.general.data(forPasteboardType:dataUTI, inItemSet:itemSet)![0] as! Data
```

Comment:
Another nit: 

Whenever I see a `!` I get nervous. Is this safe? 

Can we say that it's always true that whenever a `pasteboardUTISet.contains(dataUTI)` that `UIPasteboard.general.data(forPasteboardType:dataUTI, inItemSet:itemSet)[0]` will be a non-nil Data?

I mean, it seems like a reasonable assumption, but since this method returns an optional anyway, I'd be more comfortable writing something conservative like:

```
guard let data = UIPasteboard.general.data(forPasteboardType:dataUTI, inItemSet:itemSet)?.firstObject as Data else { 
  Logger.warn("uh oh")
  return nil
}
return imageAttachment(data: data, dataUTI: dataUTI)
```

And avoid any chance of crashing.